### PR TITLE
RGBWW - added channel interlock for RGB vs white

### LINKED ISSUE #1180: RGBWW doesn't turn off RGB when using white levels
**Operating environment/Installation (Hass.io/Docker/pip/etc.):**

hass.io, esphome 1.15.0-dev

**ESP (ESP32/ESP8266, Board/Sonoff):**

generic esp8266 based rgbww bulb

**Affected component:**

**Description of problem:**
after seeing that the code for rgbww had been altered in 1.15.0-dev compared with the version on the normal branch in hassio addons I tried it out. It works much better than the previous verison. it doesn't go far enough though as the white temperature is marred by the rgb leds remaining on when using the white channels 
I noticed in the changes between these two versions that if the colour temperature or white value is set, the rgb is defaults to 255,255,255. For my bulbs at least, the code should recognise that the RGB channels should not be used if 255,255,255 is set and thus leave them turned off. To provide flexibility for different bulbs that might need the RGB to be on, maybe a new configuration value to control this behaviour? 

**Problem-relevant YAML-configuration entries:**
```yaml
PASTE YAML FILE HERE

```

**Logs (if applicable):**

```
PASTE DEBUG LOG HERE

```

**Additional information and things you've tried:**

### PR DESCRIPTION
**Pull request by peq123**
## Description:
so that the RGB doesn't affect white colour temp interlocks RGB channels against white value 
white value is now set to either 0.0f or 1.0f so that brightness value and colour temp is now the only value controlling % of white channels. 

**Related issue (if applicable):** fixes <https://github.com/esphome/issues/issues/1180>
**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** https://github.com/esphome/esphome-docs/pull/653

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by Skaronator** [comment]
Maybe it would make sense to have this optional as a config parameter similar how the constant_brightness was implemented in #1007?

**Comment by peq123** [comment]
Honestly I haven't seen a use case to make it optional (though obviously I don't own every bulb out there). the reason why I made this pr was as follows: Google Home doesn't offer a separate white level specific for the white leds (colour temp provided but this is different), it assumes the bulbs brightness is controlled by the brightness control regardless of the leds used. The situation arose that brightness was at max but the white level at minimum and therefore the bulbs were not lit. In home assistant, this would mean extra clicks to make sure all the settings were correct after a call by Google or as I have seen several posts on home assistant forums people use HA automations or node red to address this (and to avoid extra bars). this pr addresses (and imo corrects) this behaviour. furthermore, this behaviour is how tasmota handles these types of bulbs.

**Comment by glmnet** [comment]
So with these changes you cannot turn on CW WW and RGB channels at same time?
If this is changing that I'd like to have a new option for this, we don't know how many people is using the old behavior and people naturally doesn't like changes

**Comment by peq123** [comment]
@glmnet "people naturally doesn't like changes" true true, I will add and test on my local bulbs before I re-commit

**Comment by peq123** [comment]
ok so I have added the optional flag and have tested on my bulbs. Just needs to pass Travis

**Comment by Skaronator** [comment]
Thanks, I'll give this a try tomorrow.

**Comment by peq123** [comment]
@Skaronator noticed there was a commit for the gamma fixing while I was doing this, could you check my pr contains the fixes

**Comment by Platzii** [comment]
Tested this PR with my RGBWW bulb and seems to do the job 👍

**Comment by glmnet** [comment]
So this will need a small entry in docs for `color_interlock`

**Comment by peq123** [comment]
https://github.com/esphome/esphome-docs/pull/653

File: esphome/components/light/light_color_values.h [hunk 1]
```diff
@@ -179,22 +179,30 @@ class LightColorValues {
...
-    *blue = gamma_correct(this->state_ * this->brightness_ * this->blue_, gamma);
+  void as_rgb(float *red, float *green, float *blue, float gamma = 0, bool color_interlock = false) const {
+    if (color_interlock) {
+      *red = gamma_correct(this->state_ * this->brightness_ * (1.0f - this->white_) * this->red_, gamma);
+      *green = gamma_correct(this->state_ * this->brightness_ * (1.0f - this->white_) * this->green_, gamma);
+      *blue = gamma_correct(this->state_ * this->brightness_ * (1.0f - this->white_) * this->blue_, gamma);
+    } else {
+      *red = gamma_correct(this->state_ * this->brightness_ * this->red_, gamma);
+      *green = gamma_correct(this->state_ * this->brightness_ * this->green_, gamma);
+      *blue = gamma_correct(this->state_ * this->brightness_ * this->blue_, gamma);
```
Comment (Skaronator) [inline esphome/components/light/light_color_values.h:190]: There is a lot of duplicated code above. You could simplify by applying the `this->state_ * this->brightness_ * (1.0f - this->white_)` to a variable and reuse that in the red, green and blue calculation.

The big `if` could also be minimized by just applying the logic once to the unified variable.

**Review by Skaronator — CHANGES_REQUESTED** [review]

**Comment by glmnet** [comment]
@peq123 would you like to apply the suggestion above?

**Comment by peq123** [comment]
Certainly, will start to address it this evening.

**Comment by kll** [comment]
I'm testing this on a Feit Electric A19 RGBWW that uses a combo of PWM and an SM16716 and it looks to be working exactly like I need it to. This particular bulb also has the diffuser cut off so I can see the individual LEDs and I've visually verified the interlock behavior. :+1:

**Comment by CarlosGS** [comment]
This is very helpful. Lights definitely had an undesired blue tone.

But most importantly, the bulbs were getting *extremely* toasty. I even capped all PWMs at 60% and bulbs were still getting burning hot.

I know it's "breaking", but **please consider enabling this setting by default.**

**Pull request merged**